localstack/services/s3/stream_patch.py

```diff
@@ -0,0 +1,542 @@
+        start_byte: int,
+        end_byte: int,
+    ) -> StreamedFakeKey:
+        dest_bucket = self.get_bucket(dest_bucket_name)
+        multipart = dest_bucket.multiparts[multipart_id]
+
+        src_part = self.get_object(src_bucket_name, src_key_name, version_id=src_version_id)
+
+        if start_byte is not None:
+            src_value = PartialStream(src_part.value, start_byte, end_byte)
+        else:
+            src_value = src_part.value
+        return multipart.set_part(part_id, src_value)
+
+    @patch(s3_models.S3Backend.put_object, pass_target=False)
+    def put_object(
+        self,
+        bucket_name: str,
+        key_name: str,
+        value: bytes | IO[bytes],
+        storage: Optional[str] = None,
+        etag: Optional[str] = None,
+        multipart: Optional[StreamedFakeMultipart] = None,
+        encryption: Optional[str] = None,
+        kms_key_id: Optional[str] = None,
+        bucket_key_enabled: Any = None,
+        lock_mode: Optional[str] = None,
+        lock_legal_status: Optional[str] = None,
+        lock_until: Optional[str] = None,
+        checksum_value: Optional[str] = None,
+    ) -> StreamedFakeKey:
+        key_name = clean_key_name(key_name)
+        if storage is not None and storage not in s3_models.STORAGE_CLASS:
+            raise s3_exceptions.InvalidStorageClass(storage=storage)
+
+        bucket = self.get_bucket(bucket_name)
+
+        # getting default config from bucket if not included in put request
+        if bucket.encryption:
+            bucket_key_enabled = bucket_key_enabled or bucket.encryption["Rule"].get(
+                "BucketKeyEnabled", False
+            )
+            kms_key_id = kms_key_id or bucket.encryption["Rule"][
+                "ApplyServerSideEncryptionByDefault"
+            ].get("KMSMasterKeyID")
+            encryption = (
+                encryption
+                or bucket.encryption["Rule"]["ApplyServerSideEncryptionByDefault"]["SSEAlgorithm"]
+            )
+        if isinstance(value, bytes):
+            value = BytesIO(value)
+
+        new_key = StreamedFakeKey(
+            name=key_name,
+            bucket_name=bucket_name,
+            value=value,
+            account_id=self.account_id,
+            storage=storage,
+            etag=etag,
+            is_versioned=bucket.is_versioned,
+            # AWS uses VersionId=null in both requests and responses
+            version_id=str(random.uuid4()) if bucket.is_versioned else "null",
+            multipart=multipart,
+            encryption=encryption,
+            kms_key_id=kms_key_id,
+            bucket_key_enabled=bucket_key_enabled,
+            lock_mode=lock_mode,
+            lock_legal_status=lock_legal_status,
+            lock_until=lock_until,
+            checksum_value=checksum_value,
+        )
+
+        existing_key = bucket.keys.get(key_name)
+        if existing_key:
+            existing_key.is_latest = False
+
+        bucket.keys[key_name] = new_key
```

Comment:
todo: set this back to
```suggestion
        existing_keys = bucket.keys.getlist(key_name, [])
        if bucket.is_versioned:
            keys = existing_keys + [new_key]
        else:
            keys = [new_key]
        bucket.keys.setlist(key_name, keys)
```